### DISCUSSION TITLE
Hvordan ønsker vi gjøre skjemavalidering?

### DISCUSSION THREAD

**Discussion by wkillerud** [fremtind/jokul#2595] [Annet]
tl;dr - hva tenker du om eksempelet og dokumentasjonen i løsningsforslaget?

### Behovet

[Skjemadesign under Profil](https://jokul.fremtind.no/profil/skjemadesign) har litt lite dokumentasjon om [skjemavalidering](https://jokul.fremtind.no/profil/skjemadesign#skjemavalidering). Det er flere som har ønsket seg et mer utfyllende eksempel på skjemavalidering, som det går an å ha som utgangspunkt når en skal designe eller programmere. Vi vil lese mer om hvordan skjemavalideringen skal fungere, og ha et interaktivt eksempel så det er enklere å forstå detaljene i animasjoner og tastaturnavigasjon.

### ⚠️  Disclaimer ⚠️ 

- Vi er ikke så samkjørte på tvers av løsninger at alle ar gjort det likt fram til nå 😄 
- Det er heller ikke sånn at det er et mål at ting _skal_ være likt overalt 👩‍🎨

Kort sagt: "vi gjør det ikke sånn her i dag i team X" er kjent 😜

Det jeg veldig gjerne vil ha tilbakemeldinger på er:
- Eventuelt hvorfor løsningsforslaget ikke passer i din context
- Om du mener "defaulten" vår bør være annerledes enn i forslaget, og i så fal hvorfor

### Prior art

- #2340 
- #1852

### Løsningsforslag

- #2594 

🖥️ [Klikk her for et interaktivt eksempel (opprinnelig, se Pull Request for gjeldende forslag)](https://j0iky.csb.app) ([kode](https://codesandbox.io/s/jkl-skjemavalideringseksempel-j0iky?file=/src/App.tsx))

<details><summary>Klikk her for forslag til dokumentasjon i tekstformat (opprinnelig, se Pull Request for gjeldende forslag) </summary>
<p>

## Skjemavalidering

Vi følger i stor grad det funksjonelle i [uutilsynets løsningsforslag for skjema](https://www.uutilsynet.no/wcag-standarden/skjema/38#formidle_feil_i_skjema_1). Vi skriver litt om noen av disse temaene i vår egen [oppslagsguide for universell utforming](https://jokul.fremtind.no/universell-utforming/guide) også.

Kort fortalt:

-   Vi validerer skjemaet for første gang når brukeren prøver å sende inn skjemaet eller det gjeldende steget
-   Hvert skjemafelt merkes med [statusfargen vår for feil](https://jokul.fremtind.no/profil/farger) og får en beskrivende feilmelding
-   På større skjermer lister vi også alle feilmeldingene i skjemaet i en [ErrorMessageBox](https://jokul.fremtind.no/komponenter/messagebox) i starten av skjemaet
-   Vi scroller opp til denne listen automatisk og flytter fokus til første skjemafelt, eller til det første skjemafeltet med feil på små skjermer
-   Når brukeren retter feilen skjuler vi feilmeldingen med én gang begge steder, og går tilbake til standard farge
-   På større skjermer, når alle feil er rettet, fjerner vi oppsummeringen fra starten av skjemaet

### Valider skjemaet ved forsøk på innsending

Mange skjemaer er delt inn i flere sider eller steg, mens andre er korte nok til at alt er på samme side. Om et skjema har flere steg skal hvert steg valideres for seg selv. Et steg må være uten feil for at brukeren skal kunne gå videre til neste steg.

Vi viser ingen valideringsfeil før brukeren har prøvd å trykke seg videre i skjemaet.

Et unntak gjøres dersom et valg gjort i skjemaet gjør at vi vet at brukeren ikke får lov til å gå videre i flyten. I disse tilfellene viser vi en [InfoMessageBox](/komponenter/messagebox) under skjemafeltet som forklarer hvorfor brukeren ikke trenger fortsette med utfyllingen av skjemaet, og hva neste steg er for dem. Beskjeden dukker opp med en gang valget er gjort så brukeren slipper fylle ut unødvendig informasjon.

Skjermbilde av et skjema hvor vi informerer brukeren om riktig vei videre i en InfoMessageBox

### Liste med feilene i skjemaet

Et skjema kan ofte ha flere feil, og noen ganger feil som er knyttet til hverandre, gjerne kalt kryssvalidering. For å gi brukeren en oversikt over feilene lister vi opp samtlige feil i en [ErrorMessageBox](https://jokul.fremtind.no/komponenter/messagebox) som vises over skjemaet (se [FormErrorMessageBox](https://jokul.fremtind.no/komponenter/messagebox#i-skjema) for en variant med ferdige animasjoner).

Innholdet i denne meldingsboksen skal være _selve feilmeldingene_. Det holder ikke å liste _ledeteksten_ til skjemafeltene det gjelder. Innholdet blir lest opp av skjermlesere.

Vi scroller automatisk opp så toppen av meldingsboksen er synlig for brukeren. Første inputfelt under oppsummeringen får tastaturfokus, sånn at tastaturbrukere enklere kan komme i gang med rettingen av feil.

Vi gjør _ikke_ dette på små skjermer, siden denne listen potensielt da opptar hele skjermen. På små skjermer scroller vi i stedet rett til det første skjemafeltet som har en valideringsfeil.

Skjermbilde av en oppsummering av fire valideringsfeil vist i en ErrorMessageBox

### Feilmeldinger under skjemafelt

Om et skjemafelt har feil viser vi en feilmeldingstekst under feltet. Vi har valgt at valideringsteksten skal erstatte hjelpeteksten. Derfor er det veldig viktig at feilmeldingen også tar med seg informasjonen fra hjelpeteksten, i tillegg til å forklare hva som er feil.

Eksempler på riktig og feil bruk av valideringsfeil under skjemafelt

### Retting av feil

Feilmeldingene forsvinner og skjemaet går tilbake til vanlig farge med én gang feilen er rettet. Brukeren skal ikke trenge å gå videre til neste skjemafelt for at statusen skal oppdateres.

Vi fjerner feilmeldingen fra oppsummeringen på samme måte som for skjemafeltet. Den forsvinner med én gang feilen er rettet. Når siste feil er rettet skjuler vi selve meldingsboksen.

### Vertikal bevegelse

Siden høyden på oppsummeringen av feil endrer seg underveis er det en fare for at scrollposisjonen til brukeren endres.

Moderne nettlesere har støtte for [CSS Scroll Anchoring](https://drafts.csswg.org/css-scroll-anchoring) uten at vi trenger å foreta oss noe, [med unntak av Safari](https://caniuse.com/css-overflow-anchor) som ikke har støtte i skrivende stund. Gitt den relativt lave markedsandelen til Safari på desktop aksepterer vi en noe dårligere opplevelse, frem til den nettleseren også bygger støtte for CSS-spesifikasjonen.

Det vil uansett bli noe vertikal bevegelse når vi fjerner feilmeldingene fra skjemafeltene.

### Eksempel på skjemavalidering

Det er mange detaljer her, og mye å sette seg inn i. Et bilde sier mer enn tusen ord, og et interaktivt eksempel sier enda mer enn det!

Lenke til eksempelet

</p>
</details>

![Skjermbilde av dokumentasjonen beskrevet over i Jøkul-drakt](https://user-images.githubusercontent.com/1223410/145804238-8f9321d9-ff29-466b-8d40-0553e0d5f5b9.jpg)

**Comment by piofinn** [comment]
Det er jo vanskelig å være altfor uenig med forslag som kommer fra UU-tilsynet. Det jeg likevel reagerer litt på er dette:

> Vi [viser ikke oversikten over feil] på små skjermer, siden denne listen potensielt da opptar hele skjermen. På små skjermer scroller vi i stedet rett til det første skjemafeltet som har en valideringsfeil.

Ikke nødvendigvis at denne oppførselen er _feil_ – den er standard oppførsel i et skjema med `react-hook-form` – men det føles litt rart å skulle endre _logikken_ i en løsning på bakgrunn av _skjermstørrelsen_. I tillegg vil jeg si at det å fjerne oversikten på mobil kanskje kan virke litt mot sin hensikt? UU-tilsynet anbefaler oversikten for å

> sikre at brukeren oppfatter hvis noe er feil. Ved forsøk på innsending må skjermbildet oppdateres slik at feilmeldingene alltid havner inne i skjermbildet.

En liten skjerm, hvor kanskje ikke alle skjemafeltene med feil vises på skjermen samtidig, virker jo som akkurat det stedet man trenger en slik oversikt hvis man har den? I motsatt ende av skalaen er det kanskje _ikke_ så fornuftig med en slik oversikt i en kompleks rådgiverløsning med høy informasjonstetthet, der brukerne også gjerne har opplæring i systemet de bruker.

Nå gjør vi det jo klart her at dette er et utgangspunkt, ikke en fasit, men tenker det er greit å problematisere _hvorfor_ vi anbefaler dette som et utgangspunkt. Målet må vel i alle tilfeller være å lage skjemaer som er så korte og oversiktlige som mulig, og som _forebygger_ feilutfylling i størst mulig grad (uten å stikke kjepper i hjulene på brukeren).

Rent personlig/anekdotisk tror jeg at jeg ville foretrukket at fokus ble flyttet til første felt med feil som standard, og at man hadde en oversikt _ved innsendingsknappen_ med lenker til eventuelle felter med feil. Slik kommer man fort i gang med rettingen, og om man har scrollet forbi felter med feil får man en påminnelse _før_ man sender inn skjemaet på nytt – også på små skjermer.

Ellers veldig enig i alt som står om tidspunkter for validering og utforming av hjelpe- og feiltekster! ❤️

**Reply by wkillerud** [reply]
> > Vi [viser ikke oversikten over feil] på små skjermer, siden denne listen potensielt da opptar hele skjermen. På små skjermer scroller vi i stedet rett til det første skjemafeltet som har en valideringsfeil.
> 
> Ikke nødvendigvis at denne oppførselen er _feil_ – den er standard oppførsel i et skjema med `react-hook-form` – men det føles litt rart å skulle endre _logikken_ i en løsning på bakgrunn av _skjermstørrelsen_. I tillegg vil jeg si at det å fjerne oversikten på mobil kanskje kan virke litt mot sin hensikt? 
> [...]
> En liten skjerm, hvor kanskje ikke alle skjemafeltene med feil vises på skjermen samtidig, virker jo som akkurat det stedet man trenger en slik oversikt hvis man har den? I motsatt ende av skalaen er det kanskje _ikke_ så fornuftig med en slik oversikt i en kompleks rådgiverløsning med høy informasjonstetthet, der brukerne også gjerne har opplæring i systemet de bruker.

Ser den, og er ikke helt sikker på den selv. Som du er inne på så er det en forskjell fra skjema til skjema om det gir mening, så kanskje Jøkuls _baseline_ anbefaling bør være lik uavhengig av skjermstørrelse. Går an å tipse om alternative metoder også, så vi unngår at folk føler seg altfor bundet.

> Rent personlig/anekdotisk tror jeg at jeg ville foretrukket at fokus ble flyttet til første felt med feil som standard, og at man hadde en oversikt _ved innsendingsknappen_ med lenker til eventuelle felter med feil. Slik kommer man fort i gang med rettingen, og om man har scrollet forbi felter med feil får man en påminnelse _før_ man sender inn skjemaet på nytt – også på små skjermer.

Det å ha oppsummeringen nederst, men som scrolles vekk fra, mener jeg gjør den lite verdifull på små flater annet enn som en "sjekkliste" før brukeren trykker på submit-knappen. Men ved en resubmit, om det fremdeles er feil, så gjentar vi bare scrollingen opp til toppen og oppsummeringen.

Om det er sjeldent, eller første gang jeg møter et skjema setter jeg personlig pris på å få listet opp det jeg eventuelt har gjort feil _før_ jeg setter i gang å rette. Da vet jeg noe om omfanget, og om jeg har gjort noen valg andre steder i skjemaet som kanskje har påvirket at jeg havnet der jeg gjorde. Mener for eksempel Altinn gjør det på denne måten, men husker ikke – går så lenge mellom hver gang 😄

Dette illustrerer vel bare igjen at vi ikke kan komme med en regelbok som skal følges til punkt og prikke for alle skjemaer 😅

**Reply by piofinn** [reply]
> Går an å tipse om alternative metoder også, så vi unngår at folk føler seg altfor bundet.

Ja, tenker at dette kanskje er det viktigste punktet her 😊

> Dette illustrerer vel bare igjen at vi ikke kan komme med en regelbok som skal følges til punkt og prikke for alle skjemaer 😅

Hehe, ja absolutt! Var jo også litt derfor jeg hadde disclaimer om "Personlig/anekdotisk." Brukere er forskjellige, og når vi lager så forskjellige løsninger for så forskjellige grupper må man nesten tilpasse etter det 😊

**Comment by wkillerud** [comment]
Tok inn endringsforslag så vi ikke skiller på små og store flater, og åpner med et avsnitt om at dette bare er en anbefaling/et utgangspunkt, og ikke må følges til punkt og prikke.

- #2594

**Comment by wkillerud** [comment]
**Oppsummert fra forum 21.12.21**

- Vi ble enige om å gå for forslaget i første omgang slik det ligger i #2594
- @joms kom med et forslag hvor oppsummeringen av skjemafeil hadde en lenke etter hver feilmelding som ved klikk/Enter flyttet fokus til det relevante inputfeltet. Noe vi ønsker å utforske.
- Flere var usikre på om oppsummering i toppen og scrolling bør forbli en del av standarden. Ønsket å teste med oppsummering under submit-knapp. Kan være en ulempe for tastaturbrukere, men hvis oppsummeringen har lenker til skjemafeltene vil det kunne gjøre det enklere å navigere.
